crates/ra_assists/src/raw_string.rs

```diff
@@ -0,0 +1,411 @@
+use hir::db::HirDatabase;
+use ra_syntax::{ast::AstNode, ast::Literal, SyntaxText, TextRange, TextUnit};
+
+use crate::{assist_ctx::AssistBuilder, Assist, AssistCtx, AssistId};
+
+pub(crate) fn make_raw_string(mut ctx: AssistCtx<impl HirDatabase>) -> Option<Assist> {
+    let literal = ctx.node_at_offset::<Literal>()?;
+    if literal.token().kind() == ra_syntax::SyntaxKind::STRING {
+        ctx.add_action(AssistId("make_raw_string"), "make raw string", |edit| {
+            edit.target(literal.syntax().text_range());
+            edit.insert(literal.syntax().text_range().start(), "r");
+        });
+        ctx.build()
+    } else {
+        None
+    }
+}
+
+pub(crate) fn make_usual_string(mut ctx: AssistCtx<impl HirDatabase>) -> Option<Assist> {
+    let literal = ctx.node_at_offset::<Literal>()?;
+    if literal.token().kind() == ra_syntax::SyntaxKind::RAW_STRING {
+        ctx.add_action(AssistId("make_usual_string"), "make usual string", |edit| {
+            let text = literal.syntax().text();
+            let usual_start_pos = text.find_char('"').unwrap(); // we have a RAW_STRING
+            let end = literal.syntax().text_range().end();
+            let mut i = 0;
+            let mut pos = 0;
+            let mut c = text.char_at(end - TextUnit::from(i));
+            while c != Some('"') {
+                if c != None {
+                    pos += 1;
+                }
+                i += 1;
+                c = text.char_at(end - TextUnit::from(i));
+            }
+
+            edit.target(literal.syntax().text_range());
+            edit.delete(TextRange::from_to(
+                literal.syntax().text_range().start(),
+                literal.syntax().text_range().start() + usual_start_pos,
+            ));
+            edit.delete(TextRange::from_to(
+                literal.syntax().text_range().end() - TextUnit::from(pos),
+                literal.syntax().text_range().end(),
+            ));
+            // parse inside string to escape `"`
+            let start_of_inside = usual_start_pos + TextUnit::from(1);
+            let end_of_inside = text.len() - usual_start_pos - TextUnit::from(1);
+            let inside_str = text.slice(TextRange::from_to(start_of_inside, end_of_inside));
+            escape_double_quote(
+                edit,
+                &inside_str,
+                literal.syntax().text_range().start() + start_of_inside,
+            );
+        });
+        ctx.build()
+    } else {
+        None
+    }
+}
+
+pub(crate) fn add_hash(mut ctx: AssistCtx<impl HirDatabase>) -> Option<Assist> {
+    let literal = ctx.node_at_offset::<Literal>()?;
+    if literal.token().kind() == ra_syntax::SyntaxKind::RAW_STRING {
+        ctx.add_action(AssistId("add_hash"), "add hash to raw string", |edit| {
+            edit.target(literal.syntax().text_range());
+            edit.insert(literal.syntax().text_range().start() + TextUnit::from(1), "#");

```

Comment:
```suggestion
            edit.insert(literal.syntax().text_range().start() + TextUnit::of_char('r'), "#");
```

this is more readable, because it explaines what we are skippping over